add full url support and timestamps

BrooklinJazz opened the pull request:
Adds full URL support and timestamp support.

![image](https://user-images.githubusercontent.com/14877564/178842600-85353ae6-ba2a-4bb3-82f7-e5ddcb798a76.png)

This works for shortened URLs `https://youtu.be/2OHFgjuy3DI` and timestamps using `?t=`.

I thought this would be a useful feature! I would appreciate it for DockYard Academy. Let me know if you think this would be useful to include or if you'd like any changes to be made.

Completed with the help of @auranykh

BrooklinJazz committed:
add full url support and timestamps

jonatanklosko commented on lib/kino/youtube.ex:
```diff
@@ -15,10 +15,29 @@ defmodule Kino.YouTube do
+      video_id ->
+        video_id
+    end
+  end
```
It would be less strict and simpler to match on the parsed url, like this

```elixir
uri = URI.parse(url)
query = URI.decode_query(uri.query || "")

video_id =
  case {uri, query} do
    {%{host: "www.youtube.com", path: "/watch"}, %{"v" => id}} -> id
    {%{host: "youtu.be", path: "/" <> id}, %{}} -> id
    _ -> raise ArgumentError, "expected a YouTube URL, got: #{inspect(url)}"
  end

time = if query["t"], do: String.to_integer(query["t"]), else: 0

{video_id, time}
```

And then we pass time downstream and interpolate in the iframe URL directly, wdyt?

BrooklinJazz replied:
I like this approach a lot better! I'll add it in once I have some time!

jonatanklosko commented on notebooks/examples.livemd:
````diff
@@ -19,6 +19,24 @@ end of YouTube URL, as in `https://www.youtube.com/watch?v={id}` or
 
+You may also use the full URL instead.
+
+```elixir
+Kino.YouTube.new("https://www.youtube.com/watch?v=2OHFgjuy3DI")
+```
+
+Or a shortened URL.
+
+```elixir
+Kino.YouTube.new("https://youtu.be/2OHFgjuy3DI")
+```
+
+Timestamps are also supported using `?t=`.
+
+```elixir
+Kino.YouTube.new("https://youtu.be/2OHFgjuy3DI?t=3600")
+```
````
I think for the notebook, a single video is enough, let's move those examples to the docs instead :)

BrooklinJazz replied:
Sounds good 👍

jonatanklosko reviewed (commented):
Hey @BrooklinJazz! Sounds like a good idea to me!

My first thought was to support a `:time` option, but thinking about this it seems more practical to just paste the URL. Do you need backward compatibility for your use case, or is it fine if we always require URL and drop support for video id?

Sidenote: since this repo is a sort of playground, it's best if you include the git hash in when adding as a dependency :)

BrooklinJazz commented:
> Hey @BrooklinJazz! Sounds like a good idea to me!
> 
> My first thought was to support a `:time` option, but thinking about this it seems more practical to just paste the URL. Do you need backward compatibility for your use case, or is it fine if we always require URL and drop support for video id?
> 
> Sidenote: since this repo is a sort of playground, it's best if you include the git hash in when adding as a dependency :)

Glad you like it! I'm fine with dropping support. Easy enough to find and replace my current uses.

BrooklinJazz committed:
Refactor video URL to use URI parser and decode_query

BrooklinJazz commented:
I've implemented your suggested changes. 

Please let me know if this matches your vision or if there are any other changes you would like me to make! 🙏I appreciate your feedback!

The tests no longer apply, so I've removed them. I'm not clear on how to test Kino effectively, so please let me know if there are any cases I should include.

Thank you @jonatanklosko

jonatanklosko commented on lib/kino/youtube.ex:
```diff
@@ -13,21 +13,51 @@ defmodule Kino.YouTube do
+
+  ## Examples
+
+    Provide the video URL in either `https://www.youtube.com/watch?v={id}` or
+    `https://youtu.be/{id}` format.
+
+        Kino.YouTube.new("https://www.youtu.be/2OHFgjuy3DI")
+
+        Kino.YouTube.new("https://www.youtube.com/watch?v=2OHFgjuy3DI")
+
+    Optionally provide a timestamp using the `t` query parameter.
+
+        Kino.YouTube.new("https://www.youtube.com/watch?v=2OHFgjuy3DI?t=3600")
```
```suggestion
  You can pass a regular YouTube URL such as `https://www.youtube.com/watch?v={id}`,
  or the short version `https://youtu.be/{id}`.

      Kino.YouTube.new("https://www.youtu.be/2OHFgjuy3DI")

      Kino.YouTube.new("https://www.youtube.com/watch?v=2OHFgjuy3DI")

  Timestamps are also supported via the `t` query parameter.

      Kino.YouTube.new("https://www.youtube.com/watch?v=2OHFgjuy3DI&t=3600")

```

jonatanklosko commented on notebooks/examples.livemd:
```diff
@@ -5,18 +5,17 @@
 Mix.install([
   {:kino, "~> 0.5.1"},
-  {:kino_lab, "~> 0.1.0-dev", github: "jonatanklosko/kino_lab"}
+  # {:kino_lab, "~> 0.1.0-dev", github: "jonatanklosko/kino_lab"}
+  {:kino_lab, path: "./"}
```
```suggestion
  {:kino_lab, "~> 0.1.0-dev", github: "jonatanklosko/kino_lab"}
```

jonatanklosko committed:
Apply suggestions from code review

jonatanklosko reviewed (approved):
Perfect, thank you!

> The tests no longer apply, so I've removed them. I'm not clear on how to test Kino effectively, so please let me know if there are any cases I should include.

Yeah, sounds good! Kino has APIs for testing the interactive `Kino.JS.Live` components, but for static `Kino.JS` there's not much to test since it's all html/javascript.

BrooklinJazz commented:
🥳🥳🥳🥳🥳
